dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp

```diff
@@ -0,0 +1,645 @@
+try
+{
+    auto mpp_tunnel_ptr = constructRemoteSyncTunnel();
+    mpp_tunnel_ptr->setFinishFlag(true);
+    mpp_tunnel_ptr->connect(nullptr);
+    GTEST_FAIL();
+}
+catch (Exception & e)
+{
+    GTEST_ASSERT_EQ(e.message(), "MPPTunnel has finished");
+}
+
+TEST_F(TestMPPTunnelBase, ConnectWhenConnected)
+{
+    try
+    {
+        auto mpp_tunnel_ptr = constructRemoteSyncTunnel();
+        std::unique_ptr<PacketWriter> writer_ptr = std::make_unique<MockWriter>();
+        mpp_tunnel_ptr->connect(writer_ptr.get());
+        GTEST_ASSERT_EQ(mpp_tunnel_ptr->getConnectFlag(), true);
+        mpp_tunnel_ptr->connect(writer_ptr.get());
+        GTEST_FAIL();
+    }
+    catch (Exception & e)
+    {
+        GTEST_ASSERT_EQ(e.message(), "MPPTunnel has connected");
+    }
+}
+
+TEST_F(TestMPPTunnelBase, CloseBeforeConnect)
+try
+{
+    auto mpp_tunnel_ptr = constructRemoteSyncTunnel();
+    mpp_tunnel_ptr->close("Canceled");
+    GTEST_ASSERT_EQ(mpp_tunnel_ptr->getFinishFlag(), true);
+    GTEST_ASSERT_EQ(mpp_tunnel_ptr->getConnectFlag(), false);
+}
+CATCH
+
+TEST_F(TestMPPTunnelBase, CloseAfterClose)
+try
+{
+    auto mpp_tunnel_ptr = constructRemoteSyncTunnel();
+    mpp_tunnel_ptr->close("Canceled");
+    GTEST_ASSERT_EQ(mpp_tunnel_ptr->getFinishFlag(), true);
+    mpp_tunnel_ptr->close("Canceled");
+    GTEST_ASSERT_EQ(mpp_tunnel_ptr->getFinishFlag(), true);
+}
+CATCH
+
+TEST_F(TestMPPTunnelBase, ConnectWriteClose)
+try
+{
+    auto mpp_tunnel_ptr = constructRemoteSyncTunnel();
+    std::unique_ptr<PacketWriter> writer_ptr = std::make_unique<MockWriter>();
+    mpp_tunnel_ptr->connect(writer_ptr.get());
+    GTEST_ASSERT_EQ(mpp_tunnel_ptr->getConnectFlag(), true);
+    std::unique_ptr<mpp::MPPDataPacket> data_packet_ptr = std::make_unique<mpp::MPPDataPacket>();
+    data_packet_ptr->set_data("First");
+    mpp_tunnel_ptr->write(*data_packet_ptr);
+    std::this_thread::sleep_for(std::chrono::milliseconds(1000)); // wait for sendJob thread to write data
+    mpp_tunnel_ptr->close("Canceled");
+    GTEST_ASSERT_EQ(mpp_tunnel_ptr->getFinishFlag(), true);
+    GTEST_ASSERT_EQ(dynamic_cast<MockWriter *>(writer_ptr.get())->write_packet_vec.size(), 2); //Second for err msg
+    GTEST_ASSERT_EQ(dynamic_cast<MockWriter *>(writer_ptr.get())->write_packet_vec[0], "First");
+}
+CATCH
+
+TEST_F(TestMPPTunnelBase, ConnectWriteWithCloseFlag)
+try
+{
+    auto mpp_tunnel_ptr = constructRemoteSyncTunnel();
+    std::unique_ptr<PacketWriter> writer_ptr = std::make_unique<MockWriter>();
+    mpp_tunnel_ptr->connect(writer_ptr.get());
+    GTEST_ASSERT_EQ(mpp_tunnel_ptr->getConnectFlag(), true);
+    std::unique_ptr<mpp::MPPDataPacket> data_packet_ptr = std::make_unique<mpp::MPPDataPacket>();
+    data_packet_ptr->set_data("First");
+    mpp_tunnel_ptr->write(*data_packet_ptr, true);
+    std::this_thread::sleep_for(std::chrono::milliseconds(1000)); // wait for sendJob thread to write data
+    mpp_tunnel_ptr->waitForFinish();
```

Comment:
If `waitForFinish` is called, then there is no need to sleep for the sendJob thread?